Adding examples of CCAAS and support into the test-network (#560)
hyperledger/fabric-samples: TEMP_CCAAS_TUTORIAL.md

Chaincode-as-service requires you to orchestrate the build and deployment phase yourself. Whilst this is an additional step, it gives control back. The Peer still requires a 'chaincode package' to be installed. In this case this doesn't contain code, but the information about where the chaincode is hosted. (Hostname,Port,TLS config etc)

## Fabric v2.4.1 Improvements

We need to use the latest 2.4.1 release as this contains some improvements to make this process easier. The core functionality is available in earlier releases but requires more configuration.

- The docker image for the peer contains a builder for chaincode-as-a-service preconfigured. This is named 'ccaasbuilder'. This removes the need to build your own external builder and repackage and configure the peer
- The `ccaasbuilder` applications are included in the binary tgz archive download for use in other circumstances. The `sampleconfig/core.yaml` is updated as well to refer to 'ccaasbuilder'

> - The 2.4.1 Java Chaincode release has been updated to remove the need to write a custom bootstrap main class. (similar  to the NodeJS Chaincode. It is intended that this will be added to the go chaincode as well.

Reviewer: Parenthesis!

## End-to-end with the the test-network

The `test-network` and some of the chaincodes have been updated to support running chaincode-as-a-service. The commands below assume that you've got the latest fabric-samples cloned, along with the latest Fabric docker images. 

It's useful to have two terminal windows open, one for starting the Fabric Network, and a second for monitoring all the docker containers.

In your 'monitoring' window, run this to watch all activity from the all the docker containers on the `fabric_test` network

```bash
# from the fabric-samples repo
./test-network/monitordocker.sh

> ```

Reviewer: You must start test-network before monitordocker.sh will succeed.
Also, cd to test-network first to be consistent with the below command.

In the 'Fabric Network' window, start the test network

```bash
cd test-network
./network.sh up createChannel
```

You can run other variants of this command, eg to use CouchDB or CAs, without affecting the '-as-a-service' feature. The three keys steps are:

- Build a docker image of the contract. Both `/asset-transfer-basic/chaincode-typescript` and `/asset-transfer-basic/chaincode-java` have been updated with Dockerfiles
- Install, Approve and Commit a chaincode definition. This is unchanged, but the chaincode package contains connection inforamtion, not code
- Start the docker container(s) containing the contract

Note that the order listed isn't mandatory. The key thing is that the containers are runing before the first transaction is set by the peer.

This sequence can be run as follows
```
./network.sh deployCCAAS  -ccn basicts -ccp ../asset-transfer-basic/chaincode-typescript 

> ```

Reviewer: This step failed for me. When building docker image I get this error:

```
Step 5/20 : RUN npm ci && npm run package
 ---> Running in 549576af86e6
npm ERR! The `npm ci` command can only install with an existing package-lock.json or
npm ERR! npm-shrinkwrap.json with lockfileVersion >= 1. Run an install with npm@5 or
npm ERR! later to generate a package-lock.json file, then try again.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2021-12-15T02_45_29_696Z-debug.log
The command '/bin/sh -c npm ci && npm run package' returned a non-zero code: 1
Docker buid of chaincode-as-a-service container failed
Deploying chaincode-as-a-service failed
```

Author: I think the shrinkwrap file wasn't copied in.. it is now. (I hope)

# Running Chaincode as Service with the Test Network

> The chaincode-as-a-service feature is a very useful and practical way to run 'Smart Contracts'. Traditionally the Fabric Peer has taken on the role of orchestrating the complete lifecycle of the chaincode. It required access to the Docker Daemon to create images, and start containers. Java, NodeJS and Go where explicitly known to the peer including how they should be built and started. 

Reviewer: ```suggestion
The chaincode-as-a-service feature is a very useful and practical way to run 'Smart Contracts'. Traditionally the Fabric Peer has taken on the role of orchestrating the complete lifecycle of the chaincode. It required access to the Docker Daemon to create images, and start containers. Java, NodeJS and Go chaincode frameworks where explicitly known to the peer including how they should be built and started. 
```

> As a result this makes it very hard to deploy into K8S style environments, or to run in any form of debug mode. Plus the code is being rebuilt by the peer so production deployments can have some degree of uncertainty about what dependencies have been pulled in.

Reviewer: ```suggestion
As a result this makes it very hard to deploy into Kubernetes style environments, or to run in any form of debug mode. Additionally, the chaincode image is being rebuilt by the peer, therefore there is some degree of uncertainty about what dependencies have been pulled in.
```

> Chaincode-as-service requires you to orchestrate the build and deployment phase yourself. Whilst this is an additional step, it gives control back. The Peer stills requires a 'chaincode package' to be installed. In this case this doesn't contain code, but the information about where the chaincode is hosted. (Hostname,Port,TLS config etc)

Reviewer 2: still

Author: thank you @kingworm 

There will certainly be more spelling mistakes I think... still working on it :)

## Prereqs

> We need to use the latest 2.4.1 release as this contains some improvements to make this process easier. The core functionality is available in earlier releases, but isn't quite as easy to use. 

Reviewer: Be specific about the improvement... the builder programs are included in the peer image.

- ... clone samples repo etc.. TBC

## Start the test network

It's useful to have two terminal windows open, one for starting the Fabric Network, and a second for monitoring all the docker containers.

In your 'monitoring' window, run this

```bash
# from the fabric-samples repo
./test-network/monitordocker.sh
```

> All the containers on the `fabric-test` network will be monitored; remember to terminate this as it starts a docker image on the `fabric-test` network.

Reviewer: Terminate when? how?

In the 'Fabric Network' window, start the test network

```bash
cd test-network
./network.sh up createChannel
```

> Once this has started, you can run the 'Chaincode-as-a-service'. This will do the following.

Reviewer: ```suggestion
Once this has started, you can run the 'Chaincode-as-a-service' using the `deployCCAAS` command. This will do the following.
```

> - Build a docker container of the contract, in this example `/asset-transfer-basic/chaincode-typescript`

Reviewer: An image or a container?

Author: image :-)

> - Install, Approve and Commit a chaincode definition. This is unchanged, but the chaincode package contains connection inforamtion, not code

Reviewer: ```suggestion
- Install, Approve and Commit a chaincode definition. This is unchanged, but the chaincode package contains connection information, not code
```

what information?

Reviewer: inforamtion!

> - Start the docker container containing the contract

Reviewer: Does the sequence of steps 2 and 3 matter?

```
./network.sh deployCCAAS  -ccn basic -ccp ../asset-transfer-basic/chaincode-typescript -cport 9999
```

> This is very similar to the `deployCC` command, it needs the name, and path. But also needs to have the port the chaincode container is going use. As each container is on the `fabric-test` network, you might wish to alter this so there are no collisions with other chaincode containers.

Reviewer: How many instances of ccaas container run? One for each peer? They all run on port 9999?
